### ISSUE TITLE
gracefully handle bad usages of test method

### ISSUE THREAD

**Issue by jamestalmage** [sindresorhus/ava#698]
Sometimes you get sloppy / distracted and end up with a half completed test:

```js
// you have some good tests:

test('foo', t => {
  // this will never be executed because of the mistakes below
});

// This throws - saying it should be a todo test:
test('fooo');

// Ooops - did you get distracted?:
test();

// Todo should not have an implementation, it is intended to document future tests/features.
test.todo('foo', t => {});
```

Instead of throwing when someone misuses the API, I think the misuse should just be reported as a failure. It should still execute the correctly written tests.

**Comment by sindresorhus** [comment]
I like the idea of being as graceful a possible. One concern though. Look at it the other way around. What if you have a lot of tests and you use one of them incorrectly. Now it will run all the tests before letting you know you have an issue and then you'll have to rerun the whole test file again. Just a thought.

**Comment by novemberborn** [comment]
#78 will likely force our hand here. We're doing #696 so we can correctly run exclusive test but I doubt we'll try and suss out these issues statically. If we don't then we'll end up in situations where we're already running a whole bunch of tests before the fault is encountered. Then why not run to completion?

**Comment by jamestalmage** [comment]
> Now it will run all the tests before letting you know you have an issue and then you'll have to rerun the whole test file again. Just a thought.

No reason we can't report it as an error as soon as it's encountered.

**Comment by novemberborn** [comment]
> No reason we can't report it as an error as soon as it's encountered.

You probably won't notice it though.

**Comment by jamestalmage** [comment]
Why not? If it increases the error count, that's noticeable immediately.

**Comment by novemberborn** [comment]
> Why not? If it increases the error count, that's noticeable immediately.

Fair enough.

**Comment by novemberborn** [comment]
This will be easiest to implement once https://github.com/avajs/ava/pull/1776 lands.

**Comment by IssueHuntBot** [comment]
@issuehunt has funded $40.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/26820798/issues/698) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/26820798/issues/698).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by sindresorhus** [comment]
Thinking more about this now, these are programmer mistakes that should be as loud and immediate as possible. Turning them into graceful failures that only surface after the full test run makes the feedback loop slower and the errors easier to overlook. It's the opposite of what we want.

A crash or an upfront error is better UX here. The goal isn't to keep running after invalid API usage; it's to make the mistake undeniable.

If you want to catch `test.todo('foo', callback)` statically, a lint rule is the right tool.

I think we should close this as won't fix.

**Comment by novemberborn** [comment]
What does the current upfront error reporting look like?

**Comment by sindresorhus** [comment]
Depends on the type of error.

---

**Config errors** (earliest, in CLI before anything runs):
- `✖ [error message]` with gray stack trace if available
- Immediate exit

**No test files found:**
- `✖ Couldn't find any files to test`
- Or if files exist but don't match filters: `✖ Based on your configuration, N test files were found, but did not match the filters:` followed by the filter patterns

**Missing AVA import:**

```
✘ No tests found in test/foo.test.js, make sure to import "ava" at the top of your test file
─
```

**Syntax errors:**

```
Uncaught exception in test/foo.test.js

SyntaxError: Unexpected token ';'

✘ test/foo.test.js exited with a non-zero exit code: 1
─

1 uncaught exception
```

**Uncaught exceptions during load** (e.g. top-level throw):

```
Uncaught exception in test/foo.test.js

test/foo.test.js:3

 2:
 3: throw new Error('top-level throw before tests');
 4:

Error: top-level throw before tests

› file://test/foo.test.js:3:7

✘ test/foo.test.js exited with a non-zero exit code: 1
─

1 uncaught exception
```

**process.exit():**

```
✘ Exiting due to process.exit() when running test/foo.test.js

Error: Unexpected process.exit()
    at file://test/foo.test.js:4:10
    ...

✘ test/foo.test.js exited with a non-zero exit code: 1
─
```

**Unhandled rejections:**

```
Unhandled rejection in test/foo.test.js

test/foo.test.js:3

 2:
 3: Promise.reject(new Error('unhandled'));
 4:

Error: unhandled

› file://test/foo.test.js:3:16

✘ test/foo.test.js exited with a non-zero exit code: 1
─

1 test passed
1 unhandled rejection
```

**Other errors:**
- `✘ Internal error when running test/foo.js` (or just `✘ Internal error`)
- `✘ Error in shared worker`

**End-of-run summary** counts:
- `N uncaught exceptions`
- `N unhandled rejections`
- Plus the usual test pass/fail/skip/todo counts

**Comment by novemberborn** [comment]
That does seem good enough.
